core/state/snapshot/generate.go

```diff
@@ -159,169 +179,619 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
+		vals = append(vals, common.CopyBytes(iter.Value()))
+	}
+	iter.Release()
+	if len(keys) > 0 {
+		last = keys[len(keys)-1]
 	}
-	// Create an account and state iterator pointing to the current generator marker
-	accTrie, err := trie.NewSecure(dl.root, dl.triedb)
+	// If the elements cover the entire storage range, then we don't need to use the prover
+	// to check validity. We can just feed the key/vals into a stacktrie and obtain the
+	// root hash
+	if len(origin) == 0 && !aborted {
+		stackTr := trie.NewStackTrie(nil)
+		for i, key := range keys {
+			stackTr.TryUpdate(key, common.CopyBytes(vals[i]))
+		}
+		if gotRoot := stackTr.Hash(); gotRoot != root {
+			// This trie needs to be regenerated
+			log.Debug("Failed to prove storage range", "origin", hexutil.Encode(origin), "root", root, "got", gotRoot)
+			return &proofResult{
+				staleKeys:   keys,
+				staleValues: vals,
+			}, nil
+		}
+		return &proofResult{
+			last:  last,
+			count: len(keys),
+		}, nil
+	}
+	// The trie was too large to handle in one iteration. Thus, we need to do
+	// use the prover
+	if origin == nil {
+		origin = common.Hash{}.Bytes()
+	}
+	tr, err := trie.New(root, dl.triedb)
+	if err != nil { // This error is special, it needs to abort the whole procedure
+		return nil, fmt.Errorf("trie %#x is missing", root)
+	}
+	// Generate the Merkle proofs for the first and last element
+	proof := rawdb.NewMemoryDatabase()
+	if err := tr.Prove(origin, 0, proof); err != nil {
+		log.Debug("Failed to prove storage range", "origin", hexutil.Encode(origin))
+		return &proofResult{
+			staleKeys:   keys,
+			staleValues: vals,
+			tr:          tr,
+		}, nil
+	}
+	if last != nil {
+		if err := tr.Prove(last, 0, proof); err != nil {
+			log.Debug("Failed to prove storage range", "last", hexutil.Encode(last), "err", err)
+			return &proofResult{
+				staleKeys:   keys,
+				staleValues: vals,
+				tr:          tr,
+			}, nil
+		}
+	}
+	// Verify the segment with range prover, ensure that all flat states
+	// in this range correspond to merkle trie.
+	_, _, _, cont, err := trie.VerifyRangeProof(root, origin, last, keys, vals, proof)
 	if err != nil {
-		// The account trie is missing (GC), surf the chain until one becomes available
-		stats.Log("Trie missing, state snapshotting paused", dl.root, dl.genMarker)
+		log.Debug("Failed to verify storage range proof", "last", hexutil.Encode(last))
+		return &proofResult{
+			staleKeys:   keys,
+			staleValues: vals,
+			tr:          tr,
+		}, nil
+	}
+	// Range prover says the trie still has some elements on the right side but
+	// the snap database is exhausted, then data loss is detected.
+	if cont && !aborted {
+		// TODO (@holiman or @rlj) investigate:
+		// This is an odd case. How can this happen? It means our snap data
+		// is correct, but not compelete? Maybe we can just continue here without
+		// deleting the existing data.
+		// We definitely need a testcase for this path
```

Comment:
It's a special case that: all the snap states in the disk are iterated and verified, but there are some extra states on the right side in the trie. So the data is loss in the snap. Actually In my PR I have a testcase for it. https://github.com/ethereum/go-ethereum/pull/22465/files#diff-692791523abdbacf9c463a59f5ceffc649df9c6cc646ab1fdab0d6c1a4fa9ac8R193